lib/galaxy/files/sources/ascp_fsspec.py

```diff
@@ -0,0 +1,413 @@
+            rpath: Remote file path (can be full URL or just the path component)
+            lpath: Local destination path
+            **kwargs: Additional arguments (unused, for fsspec compatibility)
+
+        Raises:
+            MessageException: If download fails after all retries
+        """
+        last_exception = None
+
+        for attempt in range(self.max_retries):
+            try:
+                self._execute_ascp_transfer(rpath, lpath, attempt)
+                return  # Success!
+            except MessageException as e:
+                last_exception = e
+
+                # Check if error is retryable
+                if self._is_retryable_error(e) and attempt < self.max_retries - 1:
+                    delay = min(self.retry_base_delay ** (attempt + 1), self.retry_max_delay)
+                    log.warning(
+                        f"ascp transfer failed (attempt {attempt + 1}/{self.max_retries}): {e}. "
+                        f"Retrying in {delay:.1f} seconds..."
+                    )
+                    time.sleep(delay)
+                else:
+                    # Non-retryable error or last attempt
+                    raise
+
+        # Should not reach here, but just in case
+        raise MessageException(
+            f"Failed to download {rpath} after {self.max_retries} attempts. Last error: {last_exception}"
+        )
+
+    def _execute_ascp_transfer(self, rpath: str, lpath: str, attempt: int) -> None:
+        """Execute a single ascp transfer attempt.
+
+        Args:
+            rpath: Remote file path
+            lpath: Local destination path
+            attempt: Current attempt number (0-indexed)
+
+        Raises:
+            MessageException: If transfer fails
+        """
+        # Parse URL if provided, otherwise use configured host/user
+        parsed_url = self._parse_url(rpath)
+        user = parsed_url.get("user") or self.user
+        host = parsed_url.get("host") or self.host
+        port = parsed_url.get("port") or self.port
+        remote_path = parsed_url.get("path", rpath)
+
+        if not user or not host:
+            raise MessageException(
+                "User and host must be specified either in configuration or in the URL. "
+                f"Got user='{user}', host='{host}'"
+            )
+
+        # Type narrowing: after validation, user and host are guaranteed to be str
+        assert isinstance(user, str)
+        assert isinstance(host, str)
+
+        if not self.ssh_key:
+            raise MessageException("SSH key is required for ascp authentication.")
+
+        # Create temporary file for key content
+        key_fd = None
+        key_path = None
+
+        try:
+            # Create temporary file for key content
+            key_fd, key_path = tempfile.mkstemp(suffix=".key", text=True)
+
+            # Write key with proper permissions (required by ascp)
+            os.chmod(key_path, 0o600)
+            with os.fdopen(key_fd, "w") as key_file:
```

Comment:
The documentation above seemed to indicate self.ssh_key could be a path and not just content - is that wrong? I guess I am fine with putting the key contents in the config like this but I would imagine an actual path would be typical? Does a path work with this implementation - if not can you address that. And either way I think we should probably rename ssh_key_content above to ssh_key and update the docs to indicate a path is more typical. I guess that is my recommendation if we keep a single config option - but maybe it worth splitting those into two options instead.

Reply:
Ahhh... I see now and I guess this makes a lot of sense given how main is deployed. Maybe add a line to the doc comments explaining the why - something like "Often Galaxy jobs run on clusters that don't mount Galaxy's root or configuration directories, this configuration block will be copied to a job's directory but referenced key paths wouldn't be."

We also have these two configuration models - one for the template and one for the concrete configuration that gets materialized in jobs - maybe we could introduce the key in the template models and embed it when we make the configuration concrete - but this feels out of scope for this pull request and if I regularly wrote Ansible code I might look at the idea ass solving a problem at the wrong level of abstraction.